pkg/cluster/operation/upgrade.go

```diff
@@ -28,91 +25,48 @@ import (
 // Upgrade the cluster.
 func Upgrade(
 	getter ExecutorGetter,
-	cluster *spec.Specification,
+	topo spec.Topology,
 	options Options,
 ) error {
 	roleFilter := set.NewStringSet(options.Roles...)
 	nodeFilter := set.NewStringSet(options.Nodes...)
-	components := cluster.ComponentsByUpdateOrder()
+	components := topo.ComponentsByUpdateOrder()
 	components = FilterComponent(components, roleFilter)
 
-	leaderAware := set.NewStringSet(spec.ComponentPD, spec.ComponentTiKV)
-
-	timeoutOpt := &clusterutil.RetryOption{
-		Timeout: time.Second * time.Duration(options.APITimeout),
-		Delay:   time.Second * 2,
-	}
-
 	for _, component := range components {
 		instances := FilterInstance(component.Instances(), nodeFilter)
 		if len(instances) < 1 {
 			continue
 		}
 
 		// Transfer leader of evict leader if the component is TiKV/PD in non-force mode
-		if !options.Force && leaderAware.Exist(component.Name()) {
-			pdClient := api.NewPDClient(cluster.GetPDList(), 5*time.Second, nil)
-			switch component.Name() {
-			case spec.ComponentPD:
-				log.Infof("Restarting component %s", component.Name())
 
-				for _, instance := range instances {
-					leader, err := pdClient.GetLeader()
-					if err != nil {
-						return errors.Annotatef(err, "failed to get PD leader %s", instance.GetHost())
-					}
+		log.Infof("Restarting component %s", component.Name())
 
-					if len(cluster.PDServers) > 1 && leader.Name == instance.(*spec.PDInstance).Name {
-						if err := pdClient.EvictPDLeader(timeoutOpt); err != nil {
-							return errors.Annotatef(err, "failed to evict PD leader %s", instance.GetHost())
-						}
-					}
+		for _, instance := range instances {
+			var rollingInstance spec.RollingUpdateInstance
 
-					if err := stopInstance(getter, instance, options.OptTimeout); err != nil {
-						return errors.Annotatef(err, "failed to stop %s", instance.GetHost())
-					}
-					if err := startInstance(getter, instance, options.OptTimeout); err != nil {
-						return errors.Annotatef(err, "failed to start %s", instance.GetHost())
-					}
-				}
+			if !options.Force {
+				rollingInstance, _ = instance.(spec.RollingUpdateInstance)
```

Comment:
It's better to check the assertion result instead of check the nil value of `rollingInstance`.